lib/spack/env/cc

```diff
@@ -114,104 +114,118 @@ case "$command" in
 esac
 
 # Finish setting up the mode.
-if [ -z "$mode" ]; then
+if [[ -z $mode ]]; then
     mode=ccld
     for arg in "$@"; do
-        if [ "$arg" = -v -o "$arg" = -V -o "$arg" = --version -o "$arg" = -dumpversion ]; then
-            mode=vcheck
-            break
-        elif [ "$arg" = -E ]; then
-            mode=cpp
-            break
-        elif [ "$arg" = -c ]; then
-            mode=cc
-            break
-        fi
+        case "$arg" in
+            -v|-V|--version|-dumpversion)
+                mode=vcheck
+                break
+                ;;
+            -E)
+                mode=cpp
+                break
+                ;;
+            -c)
+                mode=cc
+                break
+                ;;
+            -S)
+                mode=as
+                break
+                ;;
+        esac
     done
 fi
 
 # Dump the version and exist if we're in testing mode.
-if [ "$SPACK_TEST_COMMAND" = "dump-mode" ]; then
+if [[ $SPACK_TEST_COMMAND = dump-mode ]]; then
     echo "$mode"
     exit
 fi
 
 # Check that at least one of the real commands was actually selected,
 # otherwise we don't know what to execute.
-if [ -z "$command" ]; then
+if [[ -z $command ]]; then
     die "ERROR: Compiler '$SPACK_COMPILER_SPEC' does not support compiling $language programs."
 fi
 
 # Save original command for debug logging
 input_command="$@"
+args=("$@")
 
-#
-# Now do real parsing of the command line args, trying hard to keep
-# non-rpath linker arguments in the proper order w.r.t. other command
-# line arguments.  This is important for things like groups.
-#
-includes=()
-libraries=()
-libs=()
-rpaths=()
-other_args=()
+# Dump parsed values for unit testing if asked for
+if [[ -n $SPACK_TEST_COMMAND ]]; then
```

Comment:
Yes, we are collecting them twice.

We collect them in a detailed, complicated way only for debug output. This code was there before. I was hoping to be able to remove it, but that didn't quite work. If we are willing to omit or change the debug output, then the whole if block can go away, reducing the script size significantly.

However, the existing, detailed way loses the ordering between the individual `-L` and `-l` options, so it wasn't good enough. Just to build, one also doesn't need the detailed ordering. It suffices to add the additional flags in the beginning or end, which is much simpler, and which this patch introduces.